lib/Tapable.js

```diff
@@ -84,11 +84,13 @@ Tapable.prototype.applyPlugins2 = function applyPlugins2(name, param1, param2) {
 
 Tapable.prototype.applyPluginsWaterfall = function applyPluginsWaterfall(name, init) {
 	if(!this._plugins[name]) return init;
-	var args = Array.prototype.slice.call(arguments, 2);
+	var args = Array.prototype.slice.call(arguments, 1);
```

Comment:
Basically before this PR, we'd have an arguments object with `[name, initValue, sharedArg1, sharedArg2]` and we'd say ok, `args = [sharedArg1, sharedArg2]`. Then we'd call `[current].concat(args)` to produce `initValue, sharedArg1, sharedArg2` which would get passed into `apply`.

On the next iteration we'd then want the first element of the array we'll pass into `apply` not be `initValue` but instead the return value of the previous handler. So we'd then call `[current].concat(args)`, where `current` is this desired value. This produces `[someReturnVal, sharedArg1, sharedArg2]`. So time and time again the only thing changing is the first element of the array we pass into `apply`, so we should be able to get away with re-using the same array while just updating its first element.

If you think it is convoluted or messy I can ax it from the PR